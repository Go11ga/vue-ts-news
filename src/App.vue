<template>
    <div class="layout">
        <header class="layout__header">
            <VTNSprite />

            <VTNHeader />

            <VTNControls />
        </header>

        <main class="layout__main">
            <router-view/>
        </main>

        <footer class="layout__footer">
            <VTNFooter />
        </footer>
    </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import VTNSprite from '@/components/VTNSprite.vue'
import VTNHeader from '@/components/VTNHeader.vue'
import VTNControls from '@/components/VTNControls.vue'
import VTNFooter from '@/components/VTNFooter.vue'
import { mapState, mapMutations } from 'vuex'
import { toast } from 'vue3-toastify'
import { IError } from '@/types/interfaces'

export default defineComponent({
    name: 'VTN',

    components: {
        VTNSprite,
        VTNHeader,
        VTNControls,
        VTNFooter
    },

    computed: {
        ...mapState([
            'errors'
        ]),

        getError(): IError {
            return this.errors.error
        }
    },

    methods: {
        ...mapMutations([
            'CLEAR_ERROR'
        ])
    },

    watch: {
        getError(newError) {
            if (Object.keys(newError).length !== 0) {
              
                toast(newError.msg, {
                    autoClose: 5000,
                    type: newError.type
                })

                setTimeout(() => {
                    this.CLEAR_ERROR()
                }, 1000)
            }
        } 
    }
})
</script>

<style lang="scss">
@import '@/assets/style/scss/layout.scss';
@import 'vue3-toastify/dist/index.css';
</style>
