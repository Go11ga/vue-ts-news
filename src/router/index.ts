import { createRouter, createWebHistory, RouteRecordRaw } from 'vue-router'

export const routes: Array<RouteRecordRaw> = [
    {
        path: '/',
        redirect: '/all/1'
    },
    {
        path: '/:channel/:page',
        name: 'VTNNews',
        props: true,
        component: () => import('@/views/VTNNews.vue')
    },
    {
        path: "/:catchAll(.*)",
        name: 'VTN404', 
        component: () => import('@/views/VTN404.vue')
    }
]

const router = createRouter({
    history: createWebHistory(process.env.BASE_URL),
    routes
})

export default router
