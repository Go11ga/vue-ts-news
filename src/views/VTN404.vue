<template>
    <div class="vtn-404">
        <div class="vtn-404__code">
            404
        </div>

        <div class="vtn-404__text">
            Страница не найдена
        </div>

        <router-link class="vtn-404__link" to="/">
            На главную
        </router-link>
    </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { mapMutations } from 'vuex'

export default defineComponent({
    name: 'VTN404',

    methods: {
        ...mapMutations([
            'SET_ACTIVE_CHANNEL'
        ])
    },

    mounted() {
        this.SET_ACTIVE_CHANNEL('')
    }
})
</script>
