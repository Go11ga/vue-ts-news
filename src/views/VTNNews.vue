<template>
    <div class="vtn-news">

        <VTNPreloader v-if="news.newsProcess && !news.news.length"/>

        <template v-if="news.news.length">
            <VTNNewsItem
                v-for="item, idx in news.news"
                :key="idx"
                :newsItem="item"
            />

            <VTNPagination :currentPage="+page" :totalCount="totalNewsPages" />
        </template>

        <div class="vtn-news__empty" v-if="!news.newsProcess && !news.news.length">
            Новостей нет
        </div>

    </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { mapState, mapMutations, mapActions } from 'vuex'
import VTNPreloader from '@/components/VTNPreloader.vue'
import VTNNewsItem from '@/components/VTNNewsItem.vue'
import VTNPagination from '@/components/ui/VTNPagination.vue'

export default defineComponent({
    name: 'VTNNews',

    props: {
        channel: {
            type: String,
            required: true
        },
        page: {
            type: String,
            required: true
        },
    },

    components: {
        VTNPreloader,
        VTNNewsItem,
        VTNPagination
    },

    computed: {
        ...mapState([
            'news'
        ]),

        totalNewsPages() {
            return Math.ceil(this.news.totalResults / this.news.newsPerPage)
        },

        pageRoute() {
            return {
                channel: this.channel,
                page: this.page
            }
        }
    },

    methods: {
        ...mapMutations([
            'SET_ACTIVE_CHANNEL',
            'SET_CURRENT_PAGE'
        ]),

        ...mapActions([
            'fetchNews'
        ])
    },

    mounted() {
        this.SET_ACTIVE_CHANNEL(this.channel)
        this.fetchNews()
    },

    watch: {
        pageRoute: {
            handler({ channel, page }) {
                this.SET_ACTIVE_CHANNEL(channel)
                this.SET_CURRENT_PAGE(+page)
                this.fetchNews()
            },
            deep: true
        }
    }
})
</script>
