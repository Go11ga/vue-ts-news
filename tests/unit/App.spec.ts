import { shallowMount, mount } from '@vue/test-utils'
import { createStore } from 'vuex'
import { createRouter, createWebHistory } from 'vue-router'
import App from '@/App.vue'
import VTN404 from '@/views/VTN404.vue'
import VTNNews from '@/views/VTNNews.vue'
import { routes } from '@/router/index'
import * as toastify from 'vue3-toastify'

jest.mock('vue3-toastify', () => ({
    toast: jest.fn() 
}))

jest.useFakeTimers()

describe('компонент App', () => {
    let wrapper: any

    const state = () => {
        return {
            errors: {
                error: {}
            }
        }
    }

    const store = createStore({
        state
    })

    beforeEach(() => {
        wrapper = shallowMount(App, {
            global: {
                plugins: [store],
                stubs: ['router-view']
            }
        })

        wrapper.vm.CLEAR_ERROR = jest.fn()
    })

    afterEach(() => {
        wrapper.unmount()
    })

    it('dummy name test', () => {
        expect(wrapper.vm.$options.name).toMatch('VTN')
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('computed getError', () => {
        const localThis = { 
            errors: { 
                error: { type: 'Error', msg: 'Some error' }
            }
        }

        const component: any = App

        expect(component.computed.getError.call(localThis)).toEqual({ type: 'Error', msg: 'Some error' })
    })

    it('watch getError', async () => {
        wrapper.vm.errors.error = { type: 'Error', msg: 'Some error' }

        await wrapper.vm.$nextTick()

        expect(toastify.toast).toHaveBeenCalledWith('Some error', {
            autoClose: 5000,
            type: 'Error'
        })

        jest.runAllTimers()
        expect(wrapper.vm.CLEAR_ERROR).toHaveBeenCalled()
    })
})

describe('router', () => {
    let wrapper: any

    const state = () => {
        return {
            news: {
                newsProcess: false,
                news: [{}, {}]
            },
            errors: {
                error: {}
            }
        }
    }

    const mutations = {
        SET_ACTIVE_CHANNEL: jest.fn()
    }

    const actions = {
        fetchNews: jest.fn()
    }

    const store = createStore({
        state,
        mutations,
        actions
    })

    const router = createRouter({
        history: createWebHistory(process.env.BASE_URL),
        routes
    })

    beforeEach(() => {
        wrapper = mount(App, {
            global: {
                plugins: [store, router],
                stubs: [
                    'VTNSprite',
                    'VTNHeader',
                    'VTNControls',
                    'VTNFooter',
                    'VTNPreloader',
                    'VTNNewsItem',
                    'VTNPagination',
                    'router-link'
                ]
            },
            props: {
                channel: 'all',
                page: 1
            }
        })

        wrapper.vm.CLEAR_ERROR = jest.fn()
    })

    afterEach(() => {
        wrapper.unmount()
    })

    it('VTNNews /', async () => {
        router.push('/')

        await router.isReady()

        expect(wrapper.findComponent(VTNNews).exists()).toBe(true)
    })

    it('VTNNews /all/1', async () => {
        router.push('/all/1')

        await router.isReady()

        expect(wrapper.findComponent(VTNNews).exists()).toBe(true)
    })

    it('VTN404', async () => {
        router.push('/qwerty')

        await router.isReady()

        expect(wrapper.findComponent(VTN404).exists()).toBe(true)
    })
})