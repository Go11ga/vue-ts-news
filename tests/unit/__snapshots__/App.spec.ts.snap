// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`компонент App snapshot 1`] = `
<div
  class="layout"
>
  <header
    class="layout__header"
  >
    <v-t-n-sprite-stub />
    <v-t-n-header-stub />
    <v-t-n-controls-stub />
  </header>
  <main
    class="layout__main"
  >
    <router-view-stub />
  </main>
  <footer
    class="layout__footer"
  >
    <v-t-n-footer-stub />
  </footer>
</div>
`;
